src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
 	uint32 rangeTableId = 1;
 	Oid insertRelationId = insertRte->relid;
 	Var *insertPartitionColumn = PartitionColumn(insertRelationId, rangeTableId);
-	bool partitionColumnsMatch = false;
 	Query *subquery = subqueryRte->subquery;
+	bool targetTableHasPartitionColumn = false;
 
 	foreach(targetEntryCell, query->targetList)
 	{
 		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		List *insertTargetEntryColumnList = pull_var_clause_default((Node *) targetEntry);
+		int insertTargetEntryColumnCount = list_length(insertTargetEntryColumnList);
+		Var *insertVar = NULL;
+		AttrNumber originalAttrNo = InvalidAttrNumber;
+		TargetEntry *subqeryTargetEntry = NULL;
 
-		if (IsA(targetEntry->expr, Var))
+		/*
+		 * We only consider target entries that include a single column. Note that this
+		 * is slightly different than directly checking the whether the targetEntry->expr
+		 * is a var since the var could be wrapped into an implicit/explicit casting.
+		 */
+		if (insertTargetEntryColumnCount != 1)
 		{
-			Var *insertVar = (Var *) targetEntry->expr;
-			AttrNumber originalAttrNo = get_attnum(insertRelationId,
-												   targetEntry->resname);
-			TargetEntry *subqeryTargetEntry = NULL;
+			continue;
+		}
 
-			if (originalAttrNo != insertPartitionColumn->varattno)
-			{
-				continue;
-			}
+		insertVar = (Var *) linitial(insertTargetEntryColumnList);
+		originalAttrNo = get_attnum(insertRelationId,
+									targetEntry->resname);
+
+		/* skip non-partition column on the target table */
+		if (originalAttrNo != insertPartitionColumn->varattno)
+		{
+			continue;
+		}
+
+		/* INSERT query includes the partition column */
+		targetTableHasPartitionColumn = true;
+
+		subqeryTargetEntry = list_nth(subquery->targetList,
+									  insertVar->varattno - 1);
+
+		/*
+		 * Corresponding (i.e., in the same ordinal position as the target table's
+		 * partition column) select target entry does not directly belong a table.
+		 * Evaluate its expression type and error out properly.
+		 */
+		if (subqeryTargetEntry->resorigtbl == InvalidOid)
+		{
+			char *exprTypeDefault = "an expression that is not a plain column";
+			char *errorDetailTemplate = "Subquery contains %s in the "
+										"same position of the target table's "
+										"partition column.";
+
+			Expr *selectTargetExpr = subqeryTargetEntry->expr;
+			NodeTag exprType = selectTargetExpr->type;
+			char *errorDetail = "";
+			char *exprTypeStr = "";
```

Comment:
Maybe `exprDescription` or `exprDescStr`? It's a full description phrase, not just the type.